go/test/endtoend/cluster/topo_process.go

```diff
@@ -0,0 +1,304 @@
+	for time.Now().Before(timeout) {
+		if topo.IsHealthy() {
+			return
+		}
+		select {
+		case err := <-topo.exit:
+			return fmt.Errorf("process '%s' exited prematurely (err: %s)", topo.Binary, err)
+		default:
+			time.Sleep(300 * time.Millisecond)
+		}
+	}
+
+	return fmt.Errorf("process '%s' timed out after 60s (err: %s)", topo.Binary, <-topo.exit)
+}
+
+// TearDown shutdowns the running topo service
+func (topo *TopoProcess) TearDown(Cell string, originalVtRoot string, currentRoot string, keepdata bool, topoFlavor string) error {
+
+	if topoFlavor == "zk2" {
+		cmd := "shutdown"
+		if keepdata {
+			cmd = "teardown"
+		}
+		topo.proc = exec.Command(
+			topo.Binary,
+			"-log_dir", topo.LogDirectory,
+			"-zk.cfg", fmt.Sprintf("1@%v:%s", topo.Host, topo.ZKPorts),
+			cmd,
+		)
+
+		err := topo.proc.Run()
+		if err != nil {
+			return err
+		}
+	} else {
+		if topo.proc == nil || topo.exit == nil {
+			return nil
+		}
+
+		topo.removeTopoDirectories(Cell)
+
+		// Attempt graceful shutdown with SIGTERM first
+		_ = topo.proc.Process.Signal(syscall.SIGTERM)
+
+		select {
+		case <-topo.exit:
+			topo.proc = nil
+			return nil
+
+		case <-time.After(10 * time.Second):
+			topo.proc.Process.Kill()
+			topo.proc = nil
+			return <-topo.exit
+		}
+	}
+
+	if !*keepData {
+		_ = os.RemoveAll(topo.DataDirectory)
+		_ = os.RemoveAll(currentRoot)
+	}
+	_ = os.Setenv("VTDATAROOT", originalVtRoot)
+	return nil
+}
+
+// IsHealthy function checks if topo server is up and running
+func (topo *TopoProcess) IsHealthy() bool {
+	resp, err := http.Get(topo.VerifyURL)
+	if err != nil {
+		return false
+	}
+	if resp.StatusCode == 200 {
+		return true
+	}
+	return false
+}
+
+func (topo *TopoProcess) removeTopoDirectories(Cell string) {
+	_ = topo.ManageTopoDir("rmdir", "/vitess/global")
+	_ = topo.ManageTopoDir("rmdir", "/vitess/"+Cell)
+}
+
+// ManageTopoDir creates global and zone in etcd2
+func (topo *TopoProcess) ManageTopoDir(command string, directory string) (err error) {
+	url := topo.VerifyURL + directory
+	payload := strings.NewReader(`{"dir":"true"}`)
+	if command == "mkdir" {
+		req, _ := http.NewRequest("PUT", url, payload)
+		req.Header.Add("content-type", "application/json")
+		_, err = http.DefaultClient.Do(req)
+		return err
+	} else if command == "rmdir" {
+		req, _ := http.NewRequest("DELETE", url+"?dir=true", payload)
+		_, err = http.DefaultClient.Do(req)
+		return err
+	} else {
+		return nil
+	}
+}
+
+// TopoProcessInstance returns a TopoProcess handle for a etcd sevice,
+// configured with the given Config.
+// The process must be manually started by calling setup()
+func TopoProcessInstance(port int, peerPort int, hostname string, flavor string, name string) *TopoProcess {
+	binary := "etcd"
+	if flavor == "zk2" {
+		binary = "zkctl"
+	}
+	if flavor == "consul" {
+		binary = "consul"
+	}
+
+	topo := &TopoProcess{
+		Name:     name,
+		Binary:   binary,
+		Port:     port,
+		Host:     hostname,
+		PeerPort: peerPort,
```

Comment:
do we need to store both PeerPort and PeerURL?

Reply:
@deepthi  Yes, etcd required these ports. 
Python also uses two ports but they name them as clientPort and peerPort.

Reply:
Right, we only need PeerURL in this case. I will remove the extra variable.